docs/standard/serialization/binaryformatter-migration-guide/reading-nrbf-payloads.md

```diff
@@ -0,0 +1,236 @@
+}
+else if (rootObject is SZArrayRecord<byte> arrayOfBytes)
+{
+    Console.WriteLine($"It was an array of `{arrayOfBytes.Length}`-many bytes.");
+}
+```
+
+Beside `Decode`, the `NrbfDecoder` exposes a `DecodeClassRecord` method that returns `ClassRecord` (or throws).
+
+#### ClassRecord
+
+The most important type that derives from `SerializationRecord` is `ClassRecord` which represents **all `class` and `struct` instances beside arrays and natively supported primitive types**. It allows to read all member names and values. It's recommended to read the [BinaryFormatter functionality reference](./functionality-reference.md) to understand what *member* is.
+
+The API it provides:
+
+- `MemberNames` property that gets the names of serialized members.
+- `HasMember` method that checks if member of given name was present in the payload. It was designed for handling versioning scenarios where given member could have been renamed.
+- A set of dedicated methods for retrieving primitive values of the provided member name: `GetString`, `GetBoolean`, `GetByte`, `GetSByte`, `GetChar`, `GetInt16`, `GetUInt16`, `GetInt32`, `GetUInt32`, `GetInt64`, `GetUInt64`, `GetSingle`, `GetDouble`, `GetDecimal`, `GetTimeSpan` and `GetDateTime`.
+- `GetClassRecord` and `GetArrayRecord` methods to retrieve instance of given record types.
+- `GetSerializationRecord` to retrieve any serialization record and `GetRawValue` to retrieve any serialization record or a raw primitive value.
+
+Let's see it in action:
+
+```cs
+[Serializable]
+public class Sample
+{
+    public int Integer;
+    public string? Text;
+    public byte[]? ArrayOfBytes;
+    public Sample? ClassInstance;
+}
+
+ClassRecord rootRecord = NrbfDecoder.DecodeClassRecord(payload);
+Sample output = new()
+{
+    // using the dedicated methods to read primitive values
+    Integer = rootRecord.GetInt32(nameof(Sample.Integer)),
+    Text = rootRecord.GetString(nameof(Sample.Text)),
+    // using dedicated method to read an array of bytes
+    ArrayOfBytes = ((SZArrayRecord<byte>)rootRecord.GetArrayRecord(nameof(Sample.ArrayOfBytes))).GetArray(),
+    // using GetClassRecord to read a class record
+    ClassInstance = new()
+    {
+        Text = rootRecord
+            .GetClassRecord(nameof(Sample.ClassInstance))!
+            .GetString(nameof(Sample.Text))
+    }  
+};
+```
+
+#### ArrayRecord
+
+`ArrayRecord` defines the core behavior for NRBF array records and provides a base for derived classes. It provides two properties:
+
+- `Rank` which gets the rank of the array.
+- `Lengths` which get a buffer of integers that represent the number of elements in every dimension.
+
+And one method: `GetArray`. When used for the fist time, it allocates an array and fills it with the data provided in the serialized records (in case of the natively supported primitive types like `string` or `int`) or the serialized records themselves in case of arrays of complex types.
```

Comment:
```suggestion
It also provides one method: `GetArray`. When used for the first time, it allocates an array and fills it with the data provided in the serialized records (in case of the natively supported primitive types like `string` or `int`) or the serialized records themselves (in case of arrays of complex types).
```